x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/search/SearchApplicationIndexService.java

```diff
@@ -218,22 +220,42 @@ private static String getSearchAliasName(SearchApplication app) {
      * @param listener The action listener to invoke on response/failure.
      */
     public void putSearchApplication(SearchApplication app, boolean create, ActionListener<IndexResponse> listener) {
-        createOrUpdateAlias(app, new ActionListener<>() {
-            @Override
-            public void onResponse(AcknowledgedResponse acknowledgedResponse) {
-                updateSearchApplication(app, create, listener);
-            }
+        updateSearchApplication(app, create, listener.delegateFailure((l, indexResponse) -> {
+            createOrUpdateAlias(app, new ActionListener<>() {
+                @Override
+                public void onResponse(AcknowledgedResponse acknowledgedResponse) {
+                    l.onResponse(indexResponse);
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    // Convert index not found failure from the alias API into an illegal argument
+                    if (e instanceof IndexNotFoundException) {
+                        e = new IllegalArgumentException(e.getMessage(), e);
+                    }
+                    final Exception failure = e;
 
-            @Override
-            public void onFailure(Exception e) {
-                // Convert index not found failure from the alias API into an illegal argument
-                Exception failException = e;
-                if (e instanceof IndexNotFoundException) {
-                    failException = new IllegalArgumentException(e.getMessage(), e);
+                    // Updating alias failed. If it was just created, let's remove it so it is consistent with no alias being created.
```

Comment:
This gets tricky real quick. We not only have to deal with converting exceptions in some cases, but we should also restore back the previous Search Application. That would mean to retrieve it first, which can also fail...